packages/material-ui/src/styles/withStyles.js

```diff
@@ -210,6 +213,62 @@ const withStyles = (stylesOrCreator, options = {}) => Component => {
       }
     }
 
+    getClasses() {
+      let generate = false;
+
+      if (!this.disableStylesGeneration) {
+        const sheetManager = this.sheetsManager.get(this.stylesCreatorSaved);
+        const sheetsManagerTheme = sheetManager.get(this.theme);
+        if (sheetsManagerTheme.sheet.classes !== this.lastRenderedClasses) {
+          this.lastRenderedClasses = sheetsManagerTheme.sheet.classes;
+          generate = true;
+        }
+      }
+
+      if (this.props.classes) {
+        if (this.props.classes !== this.lastClassesProp) {
+          this.lastClassesProp = this.props.classes;
+          generate = true;
+        }
+        if (generate) {
```

Comment:
Again, same logic as in the render, except now it's caching the last used classes pointer and toggling generation on if it changed.